File: scripts/visualize.py
# -*- coding: utf-8 -*-
import sys
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Komut satiri argumanlarini al
input_file = sys.argv[1]
output_file = sys.argv[2]

# Sample adini cikti dosyasinin ISMINDEN cikar
SAMPLE = os.path.basename(output_file).split('_qc_plots_python.png')[0]
if SAMPLE == "":
    SAMPLE = os.path.basename(os.path.dirname(output_file))

df = pd.read_csv(input_file)

# N50 hesapla
sorted_lengths = sorted(df['length'], reverse=True)
total_bases = sum(sorted_lengths)
cumsum = 0
n50 = 0
for l in sorted_lengths:
    cumsum += l
    if cumsum >= total_bases / 2:
        n50 = l
        break

# Kalite kategorileri
total_reads = len(df)
q10_pct = (df['mean_quality'] >= 10).sum() / total_reads * 100
q20_pct = (df['mean_quality'] >= 20).sum() / total_reads * 100
q30_pct = (df['mean_quality'] >= 30).sum() / total_reads * 100
total_yield_mb = total_bases / 1_000_000

print("=== SUMMARY STATISTICS ===")
print(f"GC Content   - Mean: {df['gc_content'].mean():.2f} | Median: {df['gc_content'].median():.2f}")
print(f"Read Length  - Mean: {df['length'].mean():.2f} | Median: {df['length'].median():.2f} | N50: {n50}")
print(f"Quality Score- Mean: {df['mean_quality'].mean():.2f} | Median: {df['mean_quality'].median():.2f}")
print(f"\nQ10 Above: %{q10_pct:.1f} | Q20 Above: %{q20_pct:.1f} | Q30 Above: %{q30_pct:.1f}")
print(f"Total Yield: {total_yield_mb:.2f} Mb")

sns.set_theme(style="whitegrid")
fig, axes = plt.subplots(2, 2, figsize=(16, 11))
fig.suptitle(f"{SAMPLE.capitalize()} - Long-read QC Report (Python)", fontsize=15, fontweight='bold')

# -----------------------------------------------------------------------------
# GRAFIK 1: READ LENGTH 
# -----------------------------------------------------------------------------
axes[0,0].hist(df['length'], bins=50, color='steelblue', edgecolor='black', alpha=0.7)
axes[0,0].set_xscale('log')
axes[0,0].axvline(df['length'].median(), color='orange', linestyle='dashed', linewidth=1.5, label='Median')
axes[0,0].axvline(n50, color='red', linestyle='dashed', linewidth=1.5, label='N50')
axes[0,0].set_title('Read Length Distribution (Log Scale)')
axes[0,0].set_xlabel('Length (bp) [log scale]')
axes[0,0].set_ylabel('Read Count')
axes[0,0].legend()
stats_text = f"Mean: {df['length'].mean():.1f}\nMedian: {df['length'].median():.1f}\nN50: {n50}\nTotal: {len(df)}"
axes[0,0].text(0.97, 0.97, stats_text, transform=axes[0,0].transAxes, fontsize=8,
               verticalalignment='top', horizontalalignment='right',
               bbox=dict(boxstyle='round', facecolor='white', alpha=0.8))

# -----------------------------------------------------------------------------
# GRAFIK 2: QUALITY SCORE 
# -----------------------------------------------------------------------------
axes[0,1].hist(df['mean_quality'], bins=50, color='coral', edgecolor='black', alpha=0.7)
axes[0,1].axvline(df['mean_quality'].median(), color='orange', linestyle='dashed', linewidth=1.5, label='Median')
axes[0,1].axvline(10, color='darkred', linestyle='dotted', linewidth=2, label='Q10')
axes[0,1].axvline(20, color='purple', linestyle='dotted', linewidth=2, label='Q20')
axes[0,1].axvline(30, color='blue', linestyle='dotted', linewidth=2, label='Q30')
axes[0,1].set_title('Mean Quality Score Distribution')
axes[0,1].set_xlabel('Mean Quality Score')
axes[0,1].set_ylabel('Read Count')
axes[0,1].legend()
stats_text = f"Mean: {df['mean_quality'].mean():.1f}\nMedian: {df['mean_quality'].median():.1f}\nQ10+: %{q10_pct:.1f}\nQ20+: %{q20_pct:.1f}\nQ30+: %{q30_pct:.1f}"
axes[0,1].text(0.03, 0.97, stats_text, transform=axes[0,1].transAxes, fontsize=8,
               verticalalignment='top', horizontalalignment='left',
               bbox=dict(boxstyle='round', facecolor='white', alpha=0.8))

# -----------------------------------------------------------------------------
# GRAFIK 3: GC CONTENT 
# -----------------------------------------------------------------------------
axes[1,0].hist(df['gc_content'], bins=50, color='seagreen', edgecolor='black', alpha=0.7)
axes[1,0].axvline(df['gc_content'].median(), color='orange', linestyle='dashed', linewidth=1.5, label='Median')
axes[1,0].axvline(52, color='purple', linestyle='dashed', linewidth=1.5, label='Expected GC (52%)')
axes[1,0].set_title('GC Content Distribution')
axes[1,0].set_xlabel('GC Content (%)')
axes[1,0].set_ylabel('Read Count')
axes[1,0].legend()
stats_text = f"Mean: {df['gc_content'].mean():.1f}%\nMedian: {df['gc_content'].median():.1f}%\nTotal: {len(df)}"
axes[1,0].text(0.03, 0.97, stats_text, transform=axes[1,0].transAxes, fontsize=8,
               verticalalignment='top', horizontalalignment='left',
               bbox=dict(boxstyle='round', facecolor='white', alpha=0.8))

# -----------------------------------------------------------------------------
# GRAFIK 4: LENGTH VS QUALITY - HEXBIN (YlGnBu skalasi)
# -----------------------------------------------------------------------------
hb = axes[1,1].hexbin(df['length'], df['mean_quality'],
                      gridsize=40, cmap='YlGnBu',
                      xscale='log', bins='log', mincnt=1, alpha=0.8)
plt.colorbar(hb, ax=axes[1,1], label='Read Count (log scale)')
axes[1,1].axhline(10, color='darkred', linestyle='dotted', linewidth=2, label='Q10')
axes[1,1].axhline(20, color='purple', linestyle='dotted', linewidth=2, label='Q20')
axes[1,1].axhline(30, color='blue', linestyle='dotted', linewidth=2, label='Q30')
axes[1,1].set_title('Read Length vs Quality (Density)')
axes[1,1].set_xlabel('Read Length (bp) [log scale]')
axes[1,1].set_ylabel('Mean Quality Score')
axes[1,1].legend(loc='lower right')

# Ultra-uzun okumayi isaretle (varsa)
max_idx = df['length'].idxmax()
max_length = df.loc[max_idx, 'length']
if max_length > 100000:  # 100kb'den buyukse
    axes[1,1].plot(max_length, df.loc[max_idx, 'mean_quality'],
                   'r*', markersize=15, label=f'Max: {max_length:,.0f} bp')
    axes[1,1].legend()

plt.tight_layout()
plt.savefig(output_file, dpi=150, bbox_inches='tight')
print(f"Plot saved as {output_file}!")

File: scripts/visualize.R
# -----------------------------------------------------------------------------
# 1. PAKETLER (en basta yuklenir)
# -----------------------------------------------------------------------------
suppressPackageStartupMessages({
  library(ggplot2)
  library(gridExtra)
  library(scales)
})

# -----------------------------------------------------------------------------
# 2. KOMUT SATIRI ARGUMANLARI VE VERI YUKLEME
# -----------------------------------------------------------------------------
args <- commandArgs(trailingOnly = TRUE)
input_file <- args[1]
output_file <- args[2]

# Sample adini cikti dosyasindan al
SAMPLE <- strsplit(basename(output_file), "_qc_plots_R.png")[[1]][1]
if (is.na(SAMPLE) || SAMPLE == "") {
  SAMPLE <- basename(dirname(output_file))
}

# VERIYI OKU
if (!file.exists(input_file)) {
  stop(paste("ERROR:", input_file, "not found!"))
}
df <- read.csv(input_file)

# -----------------------------------------------------------------------------
# 3. ISTATISTIK HESAPLAMALARI
# -----------------------------------------------------------------------------
total_reads <- nrow(df)
total_bases <- sum(df$length)
total_yield_mb <- total_bases / 1e6

# N50 hesapla
sorted_lengths <- sort(df$length, decreasing = TRUE)
cumsum_lengths <- cumsum(sorted_lengths)
n50_value <- sorted_lengths[which(cumsum_lengths >= total_bases / 2)[1]]

# Kalite kategorileri
q10_pct <- sum(df$mean_quality >= 10) / total_reads * 100
q20_pct <- sum(df$mean_quality >= 20) / total_reads * 100
q30_pct <- sum(df$mean_quality >= 30) / total_reads * 100

# -----------------------------------------------------------------------------
# 4. KONSOLA YAZDIR
# -----------------------------------------------------------------------------
cat("\n=== SUMMARY STATISTICS (R) ===\n")
cat(paste("GC Content   - Mean:", round(mean(df$gc_content), 2), "| Median:", round(median(df$gc_content), 2), "\n"))
cat(paste("Read Length  - Mean:", round(mean(df$length), 2), "| Median:", round(median(df$length), 2), "| N50:", n50_value, "\n"))
cat(paste("Quality Score- Mean:", round(mean(df$mean_quality), 2), "| Median:", round(median(df$mean_quality), 2), "\n"))
cat(paste("\nQ10 Above:", round(q10_pct, 1), "% | Q20 Above:", round(q20_pct, 1), "% | Q30 Above:", round(q30_pct, 1), "%\n"))
cat(paste("Total Yield:", round(total_yield_mb, 2), "Mb\n"))

# -----------------------------------------------------------------------------
# 5. GRAFIKLER - Python ile tutarli renkler
# -----------------------------------------------------------------------------

# --- GRAFIK 1: Read Length Density (Log Scale) ---
# Renk: steelblue (Python ile ayni)
p1 <- ggplot(df, aes(x = length)) +
  geom_density(fill = "steelblue", alpha = 0.7, color = "steelblue") +
  scale_x_log10(labels = comma) +
  geom_vline(xintercept = median(df$length), color = "orange", linetype = "dashed", linewidth = 1) +
  geom_vline(xintercept = n50_value, color = "red", linetype = "dashed", linewidth = 1) +
  labs(title = paste0(SAMPLE, " - Read Length Density (Log Scale)"),
       x = "Length (bp) [log10]", y = "Density") +
  theme_minimal() +
  annotate("text", x = Inf, y = Inf,
           label = paste0("Mean: ", round(mean(df$length), 1),
                          "\nMedian: ", round(median(df$length), 1),
                          "\nN50: ", n50_value,
                          "\nTotal: ", total_reads),
           hjust = 1.1, vjust = 1.1, size = 3, color = "black")

# --- GRAFIK 2: Quality Score Boxplot ---
p2 <- ggplot(df, aes(x = "", y = mean_quality)) +
  geom_boxplot(fill = "coral", alpha = 0.7, color = "black", outlier.size = 0.5) +
  geom_hline(yintercept = 10, color = "darkred", linetype = "dotted", linewidth = 1) +
  geom_hline(yintercept = 20, color = "purple", linetype = "dotted", linewidth = 1) +
  geom_hline(yintercept = 30, color = "blue", linetype = "dotted", linewidth = 1) +
  labs(title = paste0(SAMPLE, " - Quality Score Distribution"),
       x = "", y = "Mean Quality Score") +
  theme_minimal() +
  annotate("text", x = Inf, y = Inf,
           label = paste0("Mean: ", round(mean(df$mean_quality), 1),
                          "\nMedian: ", round(median(df$mean_quality), 1),
                          "\nQ10+: ", round(q10_pct, 1), "%",
                          "\nQ20+: ", round(q20_pct, 1), "%",
                          "\nQ30+: ", round(q30_pct, 1), "%"),
           hjust = 1.1, vjust = 1.1, size = 3, color = "black")

# --- GRAFIK 3: GC Content Violin Plot ---
p3 <- ggplot(df, aes(x = "", y = gc_content)) +
  geom_violin(fill = "seagreen", alpha = 0.7, color = "seagreen") +
  geom_hline(yintercept = median(df$gc_content), color = "orange", linetype = "dashed", linewidth = 1) +
  geom_hline(yintercept = 52, color = "purple", linetype = "dashed", linewidth = 1) +
  labs(title = paste0(SAMPLE, " - GC Content Distribution"),
       x = "", y = "GC Content (%)") +
  theme_minimal() +
  annotate("text", x = Inf, y = Inf,
           label = paste0("Mean: ", round(mean(df$gc_content), 1), "%",
                          "\nMedian: ", round(median(df$gc_content), 1), "%"),
           hjust = 1.1, vjust = 1.1, size = 3, color = "black")

# -----------------------------------------------------------------------------
# 6. GRAFIKLERI BIRLESTIR VE KAYDET
# -----------------------------------------------------------------------------
combined <- grid.arrange(p1, p2, p3, ncol = 3)
ggsave(output_file, combined, width = 18, height = 6, dpi = 150)

cat(paste("\n[SUCCESS] R plot saved as", output_file, "\n"))